test/jdk/sun/util/calendar/CalendarSystemDeadLockTest.java

```diff
@@ -0,0 +1,171 @@
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+
+/**
+ * @test
+ * @bug 8273790
+ * @summary Verify that concurrent classloading of sun.util.calendar.Gregorian and
+ * sun.util.calendar.CalendarSystem doesn't lead to a deadlock
+ * @modules java.base/sun.util.calendar:open
+ * @run main/othervm CalendarSystemDeadLockTest
+ * @run main/othervm CalendarSystemDeadLockTest
+ * @run main/othervm CalendarSystemDeadLockTest
+ * @run main/othervm CalendarSystemDeadLockTest
+ * @run main/othervm CalendarSystemDeadLockTest
+ */
+public class CalendarSystemDeadLockTest {
+
+    public static void main(final String[] args) throws Exception {
+        testConcurrentClassLoad();
+    }
+
+    /**
+     * Loads {@code sun.util.calendar.Gregorian} and {@code sun.util.calendar.CalendarSystem}
+     * and invokes {@code sun.util.calendar.CalendarSystem#getGregorianCalendar()} concurrently
+     * in a thread of their own and expects the classloading of both those classes
+     * to succeed. Additionally, after these tasks are done, calls the
+     * sun.util.calendar.CalendarSystem#getGregorianCalendar() and expects it to return a singleton
+     * instance
+     */
+    private static void testConcurrentClassLoad() throws Exception {
+        final int numTasks = 7;
+        final CountDownLatch taskTriggerLatch = new CountDownLatch(numTasks);
+        final List<Callable<?>> tasks = new ArrayList<>();
+        // add the sun.util.calendar.Gregorian and sun.util.calendar.CalendarSystem for classloading.
+        // there are main 2 classes which had a cyclic call in their static init
+        tasks.add(new ClassLoadTask("sun.util.calendar.Gregorian", taskTriggerLatch));
+        tasks.add(new ClassLoadTask("sun.util.calendar.CalendarSystem", taskTriggerLatch));
+        // add a few other classes for classloading, those which call CalendarSystem#getGregorianCalendar()
+        // or CalendarSystem#forName() during their static init
+        tasks.add(new ClassLoadTask("java.util.GregorianCalendar", taskTriggerLatch));
+        tasks.add(new ClassLoadTask("java.util.Date", taskTriggerLatch));
+        tasks.add(new ClassLoadTask("java.util.JapaneseImperialCalendar", taskTriggerLatch));
+        // add a couple of tasks which directly invoke sun.util.calendar.CalendarSystem#getGregorianCalendar()
+        tasks.add(new GetGregorianCalTask(taskTriggerLatch));
+        tasks.add(new GetGregorianCalTask(taskTriggerLatch));
+        final ExecutorService executor = Executors.newFixedThreadPool(tasks.size());
```

Comment:
Asserting `tasks.size() == numTasks` may help here.

Reply:
Yes, that makes sense. I've updated the test to add this check.